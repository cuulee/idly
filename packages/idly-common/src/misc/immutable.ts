// import { Map as ImMap } from 'immutable';

// export ImMap;
// // export const ImList = Immutable.List;
// // export const ImSet = Immutable.Set;
