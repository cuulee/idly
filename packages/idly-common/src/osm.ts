export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
