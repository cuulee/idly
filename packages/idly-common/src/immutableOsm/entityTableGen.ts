import { Entity, EntityTable, Way, Node, Relation } from '../osm';

// export function entityTableGen(t: EntityTable, entities: Entity[]) {
//   for (const e of entities) {
//     if (t.has(e.id)) continue;
//     t.set(e.id, e);
//   }
//   return t;
// }
