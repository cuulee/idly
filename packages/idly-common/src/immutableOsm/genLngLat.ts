import { recordify } from "typed-immutable-record";
import { $LngLat } from "../immutableOsm/immutableOsm";
import { LngLat } from "../osm/structures";

export function $genLngLat(l: LngLat): $LngLat {
  return recordify<LngLat, $LngLat>(l);
}
