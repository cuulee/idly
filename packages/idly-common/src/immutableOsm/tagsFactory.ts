import { Tags } from '../osm/structures';
import { $Tags } from '../immutableOsm/immutableOsm';
import { Map as $Map } from 'immutable';

export function $tagsFactory(t: Tags): $Tags {
  return $Map(t);
}
