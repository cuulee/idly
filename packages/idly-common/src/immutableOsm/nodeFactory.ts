import { $attributesGen } from './attributesGen';
import { $genLngLat } from './genLngLat';
import { $tagsFactory } from './tagsFactory';
import { $Node, $__Node } from '../immutableOsm';
import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import { Attributes, EntityId, EntityType, LngLat, Node, Tags } from '../osm';
import { recordify } from 'typed-immutable-record';

export function nodeFactory(n: Node): $Node {
  var $tags = $tagsFactory(n.tags);
  var $loc = $genLngLat(n.loc);
  var $attr = $attributesGen(n.attributes);
  return recordify<$__Node, $Node>({
    id: n.id,
    tags: $tags,
    type: EntityType.NODE,
    loc: $loc,
    attributes: $attr
  });
}
