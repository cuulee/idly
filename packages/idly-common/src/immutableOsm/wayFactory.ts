import { $attributesGen } from './attributesGen';
import { $tagsFactory } from './tagsFactory';
import { $Way, $__Way } from '../immutableOsm';
import { Attributes, EntityId, EntityType, Tags, Way } from '../osm';
import { attributesGen } from '../osm/attributesGen';
import { recordify } from 'typed-immutable-record';
import { List as $List } from 'immutable';

export function $wayFactory(way: Way): $Way {
  var $tags = $tagsFactory(way.tags);
  var $attr = $attributesGen(way.attributes);
  var $nodes = $List(way.nodes);
  return recordify<$__Way, $Way>({
    id: way.id,
    type: EntityType.WAY,
    tags: $tags,
    attributes: $attr,
    nodes: $nodes
  });
}
