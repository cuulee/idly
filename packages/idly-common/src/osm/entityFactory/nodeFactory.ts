import { lngLatFactory } from '../../geo/lngLatFactory';
import { deepFreeze } from '../../misc/deepFreeze';

import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function nodeFactory(
  {
    id,
    tags = tagsFactory(),
    loc = lngLatFactory([0, 0]),
    attributes = attributesFactory(),
  }: {
    id: EntityId;
    tags?: Tags;
    loc?: LngLat;
    attributes?: Attributes;
  },
  freeze = true
): Node {
  return deepFreeze<Node>(
    {
      attributes: attributesFactory(attributes),
      id,
      loc,
      tags: tagsFactory(tags),
      type: EntityType.NODE,
    },
    freeze
  );
}
