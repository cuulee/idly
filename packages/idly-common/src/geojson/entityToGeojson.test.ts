import { Map as ImMap } from 'immutable';

import { nodeFactory } from '../osm/nodeFactory';
import { genLngLat } from '../osm/genLngLat';
import { tagsFactory } from '../osm/tagsFactory';

import {
  nodeCombiner,
  wayToLineString,
  getCoordsFromTable,
  wayCombiner,
} from './entityToGeojson';
import { wayFactory } from '../osm/wayFactory';
import { OsmGeometry, Entity, EntityTable } from '../osm/structures';
import { entityTableGen } from '../osm/entityTableGen';

const n1 = nodeFactory({ id: 'n-1' });
const n11 = nodeFactory({
  id: 'n-1',
  loc: genLngLat({ lon: 15, lat: 10 }),
  tags: tagsFactory({ k: 'k' }),
});

describe('converts node to feat', () => {
  describe('nodeCombiner', () => {
    it('should behave...', () => {
      expect(nodeCombiner(n1, {})).toMatchSnapshot();
      expect(nodeCombiner(n11, {})).toMatchSnapshot();
    });
    it('should add computed Props', () => {
      expect(nodeCombiner(n1, { test: '1234', h: 3 })).toMatchSnapshot();
    });
  });
});

const way = wayFactory({
  id: 'w1',
  nodes: ['n1'],
  tags: tagsFactory({ highway: 'residential' }),
});

const node = nodeFactory({ id: 'n1' });

// tslint:disable:no-expression-statement object-literal-key-quotes
describe('way.test', () => {
  describe('wayToLineString', () => {
    it('should throw error when no geometry provided', () => {
      expect(() => wayToLineString(undefined as any, [[1, 2]])).toThrow(); // ts:disable-line
      expect(() => wayToLineString(OsmGeometry.POINT, [[1, 2]])).toThrow();
    });
    it('should not work when correct coords provided with wrong geometry', () => {
      expect(() =>
        wayToLineString(OsmGeometry.POINT, [[1, 2], [2, 3], [3, 0], [1, 2]]),
      ).toThrow();
    });
    it('should throw error when wrong coords provided', () => {
      expect(() =>
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 1]]),
      ).toThrow();
    });
    it('should work when correct coords provided', () => {
      expect(() =>
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]]),
      ).not.toThrow();
    });
    it('should behave...', () => {
      expect(
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]]),
      ).toMatchSnapshot();
    });
  });

  describe('getNodesFromGraph', () => {
    it('should get nodes correctly', () => {
      const n1: Entity = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      let g: EntityTable = ImMap();
      g = g.set(n1.id, n1);
      expect(getCoordsFromTable(g, ['n1'])).toMatchSnapshot();
    });
    it('should throw error if nodes not found', () => {
      const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      let g: EntityTable = ImMap();
      g = g.set(n1.id, n1);
      expect(() => getCoordsFromTable(g, ['n2'])).toThrow();
    });

    it('should work when multiple nodes are there ingraph', () => {
      const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
      const n2 = nodeFactory({ id: 'n2', loc: genLngLat([3, 4]) });
      const g = entityTableGen([n1, n2]);
      expect(getCoordsFromTable(g, ['n2'])).toMatchSnapshot();
      expect(getCoordsFromTable(g, ['n1', 'n2'])).toMatchSnapshot();
    });
  });

  describe('wayCombiner', () => {
    const n1 = nodeFactory({ id: 'n1', loc: genLngLat([1, 2]) });
    const n2 = nodeFactory({ id: 'n2', loc: genLngLat([3, 4]) });
    const n3 = nodeFactory({ id: 'n3', loc: genLngLat([1, 4]) });

    const w1 = wayFactory({
      id: 'w1',
      nodes: ['n1', 'n3'],
      tags: tagsFactory({ highway: 'residential' }),
    });

    const g = entityTableGen([n1, n2, n3, w1]);
    it('should behave...', () => {
      const result = wayCombiner(w1, g, {
        'osm_basic--geometry': OsmGeometry.LINE,
      });
      expect(result).toMatchSnapshot();
    });

    it('sets up existing props', () => {
      const result = wayCombiner(w1, g, {
        k: 'k',
        'osm_basic--geometry': OsmGeometry.LINE,
      });
      expect(result.properties).toMatchObject({
        id: 'w1',
        k: 'k',
        'osm_basic--geometry': 'line',
      });
    });
    it('throws error if no geometry is provided', () => {
      expect(() => wayCombiner(way, g, { s: 'k' } as any)).toThrow();
    });
  });
});
