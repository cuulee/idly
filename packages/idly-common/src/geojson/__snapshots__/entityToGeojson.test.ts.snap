// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts node to feat nodeCombiner should add computed Props 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "h": 3,
    "id": "n-1",
    "test": "1234",
  },
  "type": "Feature",
}
`;

exports[`converts node to feat nodeCombiner should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "id": "n-1",
  },
  "type": "Feature",
}
`;

exports[`converts node to feat nodeCombiner should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n-1",
  "properties": Object {
    "id": "n-1",
  },
  "type": "Feature",
}
`;

exports[`way.test getNodesFromGraph should get nodes correctly 1`] = `
Array [
  Array [
    1,
    2,
  ],
]
`;

exports[`way.test getNodesFromGraph should work when multiple nodes are there ingraph 1`] = `
Array [
  Array [
    3,
    4,
  ],
]
`;

exports[`way.test getNodesFromGraph should work when multiple nodes are there ingraph 2`] = `
Array [
  Array [
    1,
    2,
  ],
  Array [
    3,
    4,
  ],
]
`;

exports[`way.test wayCombiner should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        1,
        2,
      ],
      Array [
        1,
        4,
      ],
    ],
    "type": "LineString",
  },
  "id": "w1",
  "properties": Object {
    "id": "w1",
    "osm_basic--geometry": "line",
  },
  "type": "Feature",
}
`;

exports[`way.test wayToLineString should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          1,
          2,
        ],
        Array [
          2,
          3,
        ],
        Array [
          3,
          0,
        ],
        Array [
          1,
          2,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
